model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java

```diff
@@ -216,17 +225,33 @@ private Stream<V> pessimisticQueryRead(QueryParameters<M> queryParameters) {
             if (entity == null) {
                 return Stream.empty();
             }
-            boolean fulfillsQueryCriteria = mapMcb.getKeyFilter().test(keyConverter.fromString(id)) && mapMcb.getEntityFilter().test(entity);
+            K k = keyConverter.fromString(id);
+            boolean fulfillsQueryCriteria = mapMcb.getKeyFilter().test(k) && mapMcb.getEntityFilter().test(entity);
             if (!fulfillsQueryCriteria) {
                 // entity does not fulfill whole criteria, we can release lock now
-                HotRodLocksUtils.removeWithInstanceIdentifier(locksCache, getLockName(id));
+                if (LockObjectsForModification.isEnabled(session, storedEntityDescriptor.getModelTypeClass())) {
+                    HotRodLocksUtils.removeWithInstanceIdentifier(locksCache, getLockName(id));
+                    entityVersionCache.remove(k);
+                }
                 return Stream.empty();
             }
 
             return Stream.of(entity);
         }
 
-        // Criteria does not contain only one id, we need to read ids non-pessimistically and then read entities one by one pessimistically
+        // workaround if the query contains us.clientId field, in which case don't read by id => read without optimistic locking.
+        // See https://issues.redhat.com/browse/ISPN-14537
+        if (dmc.partiallyEvaluate((field,  op, arg) -> field == UserSessionModel.SearchableFields.CLIENT_ID).toString().contains("__TRUE__")) {
```

Comment:
This workaround can be also removed when we can project the object version within the query. Can you please create a follow-up issue for that? it should be depending on ISPN 14.0.8 - https://issues.redhat.com/browse/ISPN-14478

Reply:
Done https://github.com/keycloak/keycloak/issues/19518